examples/ros/src/smarts_ros/scripts/ros_driver.py

```diff
@@ -0,0 +1,523 @@
+            )
+        mission = Mission(
+            start=Start.from_pose(Pose.from_ros(ros_agent_spec.start_pose)),
+            goal=PositionalGoal(
+                (
+                    ros_agent_spec.end_pose.position.x,
+                    ros_agent_spec.end_pose.position.y,
+                ),
+                ros_agent_spec.veh_length,
+            ),
+            # via=TODO(ros_agent_spec.vias),
+            # TODO:  how to prevent them from spawning on top of another existing vehicle? (see how it's done in SUMO traffic)
+            entry_tactic=default_entry_tactic(ros_agent_spec.start_speed),
+            vehicle_spec=VehicleSpec(
+                veh_id=f"veh_for_agent_{ros_agent_spec.agent_id}",
+                veh_config_type=self._decode_vehicle_type(ros_agent_spec.veh_type),
+                dimensions=BoundingBox(
+                    ros_agent_spec.veh_length,
+                    ros_agent_spec.veh_width,
+                    ros_agent_spec.veh_height,
+                ),
+            ),
+        )
+        with self._control_lock:
+            if (
+                ros_agent_spec.agent_id in self._agents
+                or ros_agent_spec.agent_id in self._agents_to_add
+            ):
+                rospy.logwarn(
+                    f"trying to add new agent with existing agent_id '{ros_agent_spec.agent_id}'.  ignoring."
+                )
+                return
+            self._agents_to_add[ros_agent_spec.agent_id] = (agent_spec, mission)
+
+    @staticmethod
+    def _get_nested_attr(obj: Any, dotname: str) -> Any:
+        props = dotname.split(".")
+        for prop in props:
+            obj = getattr(obj, prop)
+        return obj
+
+    @staticmethod
+    def _extrapolate_to_now(
+        state_name: str, states: Sequence[float], veh_id: str, staleness: float
+    ) -> np.ndarray:
+        # Here, we just do some straightforward/basic smoothing using a moving average,
+        # and then extrapolate to the current time.
+        dt = 0.0
+        last_time = None
+        avg = np.array((0.0, 0.0, 0.0, 0.0))
+        for state in states:
+            for entity in state.entities:
+                if entity.entity_id != veh_id:
+                    continue
+                vector = ROSDriver._get_nested_attr(entity, state_name)
+                assert vector
+                if hasattr(vector, "w"):
+                    vector = np.array((vector.x, vector.y, vector.z, vector.w))
+                else:
+                    vector = np.array((vector.x, vector.y, vector.z))
+                avg[: len(vector)] += vector
+                stamp = state.header.stamp.to_sec()
+                if last_time:
+                    dt += stamp - last_time
+                last_time = stamp
+                last_vect = vector
+                break
+        avg /= len(states)
```

Comment:
I have some concerns that averaging quaternions takes a few more steps here since quaternions are ambiguous with their negation `q == -q` so adding for the purposes of averaging without accounting for the 2:1 mapping can cause loss of rotation.

https://stackoverflow.com/questions/12374087/average-of-multiple-quaternions

Reply:
Good catch!  But rather than do it with the approach suggested there, I was considering converting each to a heading angle, averaging those, and then converting it back to a quaternion.  Do you think that'd be ok here?

Reply:
That could work as well. Our heading class accounts for the `-180deg == 180deg` issue already.